sympy/solvers/tests/test_inequalities.py

```diff
@@ -486,3 +491,102 @@ def test__pt():
     assert _pt(x, oo) == _pt(oo, x) == x + 1
     assert _pt(x, -oo) == _pt(-oo, x) == x - 1
     raises(ValueError, lambda: _pt(Dummy('i', infinite=True), S.One))
+
+
+def test_linear_programming():
+    seed(47)
+    r1 = y+2*z <= 3
+    r2 = -x - 3*z <= -2
+    r3 = 2*x + y + 7*z <= 5
+    optimum, argmax, argmax_dual = linear_programming([r1, r2, r3], -x-y-5*z )
+    assert optimum == -2
+    assert argmax == [2, 0, 0]
+    assert argmax_dual == [0, 1, 0]
+
+    r1 = x - y + 2*z <= 3
+    r2 = -x + 2*y - 3*z <= -2
+    r3 = 2*x + y - 7*z <= -5
+    optimum, argmax, argmax_dual = linear_programming([r1, r2, r3], -x-y-5*z )
+    assert optimum == Rational(-25, 7)
+    assert argmax == [0, 0, Rational(5,7)]
+    assert argmax_dual == [0, 0, Rational(5,7)]
+
+    r1 = x - y + 2*z <= -4
+    r2 = -x + 2*y - 3*z <= 8
+    r3 = 2*x + y - 7*z <= 10
+    optimum, argmax, argmax_dual = linear_programming([r1, r2, r3], -x-y-5*z )
+    assert optimum == -4
+    assert argmax == [0, 4, 0]
+    assert argmax_dual == [1, 0, 0]
+
+    # Section 4 Problem 1 from http://web.tecnico.ulisboa.pt/mcasquilho/acad/or/ftp/FergusonUCLA_LP.pdf
+    x1, x2, x3, x4 = symbols('x1 x2 x3 x4')
+    r1 = x1 - x2 - 2*x3 - x4 <= 4
+    r2 = 2*x1 + x3 -4*x4 <= 2
+    r3 = -2*x1 + x2 + x4 <= 1
+    optimum, argmax, argmax_dual = linear_programming([r1, r2, r3], x1 - 2*x2 - 3*x3 - x4)
+    assert optimum == 4
+    assert argmax == [7, 0, 0, 3]
+    assert argmax_dual == [1, 0, 0]
+
+    # equality
+    r1 = Eq(x,y)
+    r2 = Eq(y,z)
+    r3 = z <= 3
+    optimum, argmax, argmax_dual = linear_programming([r1, r2, r3], x)
+    assert optimum == 3
+    assert argmax == [3, 3, 3]
+    assert argmax_dual == [1, 0, 1, 0, 1]
+
+    # Binary predicate
+    r1 = Q.ge(x, y)
+    r2 = Q.le(x, 4)
+    optimum, argmax, argmax_dual = linear_programming([r1, r2], x+y)
+    assert optimum == 8
+    assert argmax == [4, 4]
+    assert argmax_dual == [1, 2]
+
+    # input contains Floats
+    r1 = x - y + 2.0*z <= -4
+    r2 = -x + 2*y - 3.0*z <= 8
+    r3 = 2*x + y - 7*z <= 10
+    optimum, argmax, argmax_dual = linear_programming([r1, r2, r3], -x-y-5*z )
+    assert optimum == -4
+    assert argmax == [0, 4, 0]
+    assert argmax_dual == [1, 0, 0]
+
+    # input contains symbolic expression (sqrt(2) for example)
+    r1 = x - y + sqrt(2)*z <= -4
```

Comment:
There needs to be some definition of what are the inputs that this method can handle. I don't think that arbitrary symbolic expressions can be handled so there should be some checking for this.

I would be happier with just restricting the method to Rational or Float if it is not clear what is the full class of expressions that can be handled. I think that the algorithm requires being able to solve the zero equivalence problem which is not trivial for arbitrary expressions.

Reply:
Would it be possible to throw an error if it takes too long to compare two expressions?